Addresses #279 configurable gateway credentials.

alexwillingham opened the pull request:
The changes address issue #279 by making apig endpoint credentials optionally configurable in s-function.json

Due to the resolution of https://github.com/awslabs/aws-apigateway-swagger-importer/issues/41 which was referenced in a comment, it seems the project is ready for this?

ac360 commented:
@alexwillingham Ryan closed awslabs/aws-apigateway-importer#41 but I don't think the issue has been resolved yet, and the performance difference is severe.  I'm not sure I want to enable this feature and have people be disappointed by it again.  What do you think?

alexwillingham commented:
@ac360 Thanks for checking in on this!

I haven't noticed the slowdown myself, but perhaps I'm just not working with enough endpoints yet for it to be noticeable.  

It seems to me that by making it configurable, Serverless can support more use cases, while anyone who wishes to avoid a slowdown can continue to not provide a "credentials" value.  So long as it's optional, I don't see the trouble.

Speaking of optional, when I made this PR for JAWS initially, I added the credentials property to the awsm.json template.  When making the changes for Serverless, it think I saw there were other optional properties not in the template, so I left out credentials.  I'd be happy to add it to the template if you'd like. :-)

ac360 commented:
@alexwillingham Ok, let's add this, but I'd like to keep the comments in the code.  The latency was ~500ms so it was a big pain-point for some people and I don't want to lose track of this issue.  Could you refactor your PR to keep some of the existing comments in, or better describe the issue?  Also, if you can add the credentials property to the templates, that would be very helpful.

alexwillingham commented:
@ac360 Added credentials property to template.  Included comment warning of potential slowdown.

alexwillingham commented:
resolved conflicts since the breaking changes of 0.1.0

fuyu commented:
I'd like to endorse this option to get merged into base branch.

alexwillingham commented:
@ac360 I've rebased again and tested everything works against 0.4.2.  Works as expected:

![image](https://cloud.githubusercontent.com/assets/3759688/13688366/6fd5d7ac-e6ef-11e5-942e-31c9b77eca08.png)
![image](https://cloud.githubusercontent.com/assets/3759688/13688287/dd4dff0e-e6ee-11e5-8eda-d2539b82b799.png)

I decided to do some simple benchmarking to investigate any difference in endpoint deploy time with and without this PR's changes.  I created a 50 lambda / 50 endpoint sls project running 'sls function create' 50 times in the same component.

I ran ```date && sls endpoint deploy --debug && date``` with the code from this PR against the 50 endpoint component after changing authorizationType and adding credentials to the s-function.json for each endpoint:
```"authorizationType": "AWS_IAM",  
"credentials": "arn:aws:iam::*:user/*",  ```
https://gist.github.com/alexwillingham/c4cdef720a872e0a0f11

Then I changed authorizationType back to "none", removed "credentials" from each s-function.json, and ran the same function with 0.4.2 base against the 50 endpoint component:
https://gist.github.com/alexwillingham/e6ad917f8baab2c73b95

Unfortunately, I found that they both crashed before deploying all 50 endpoints as you'll see in the Gists!  I'll be happy to log an issue documenting it if it's not an already known bug.  

I posted here anyways, because in comparing the times shown by the --debug flag, I don't see that they're noticeably different.  I'll be happy to repeat the test if you guys can get me to where I can deploy all 50 endpoints without crashing.  Regardless, I think it's safe to merge in these changes; Let me know if I can do anything else to help.

Thanks!

alexwillingham commented:
Well that was silly of me... I understand from discussion in Gitter that the latency issue arises in invoking the endpoint, not deployment. I look forward to profiling the latency with/without these changes.

ac360 commented:
@alexwillingham I've just added this in separately to the master branch.  I don't believe the API Gateway team has fixed this.  But let us know if we're wrong!
